src/python/grpcio/grpc/_cython/_cygrpc/aio/grpc_aio.pyx.pxi

```diff
@@ -13,25 +13,97 @@
+
+        # Anchors the event loop that the gRPC library going to use.
+        _grpc_aio_loop = asyncio.get_event_loop()
+        _event_loop_thread_ident = threading.current_thread().ident
+
+        if grpc_aio_engine is AsyncIOEngine.CUSTOM_IO_MANAGER:
+            # Activates asyncio IO manager.
+            # NOTE(lidiz) Custom IO manager must be activated before the first
+            # `grpc_init()`. Otherwise, some special configurations in Core won't
+            # pick up the change, and resulted in SEGFAULT or ABORT.
+            install_asyncio_iomgr()
+
+            # TODO(https://github.com/grpc/grpc/issues/22244) we need a the
+            # grpc_shutdown_blocking() counterpart for this call. Otherwise, the gRPC
+            # library won't shutdown cleanly.
+            grpc_init()
+
+            # Timers are triggered by the Asyncio loop. We disable
+            # the background thread that is being used by the native
+            # gRPC iomgr.
+            grpc_timer_manager_set_threading(False)
+
+            # gRPC callbaks are executed within the same thread used by the Asyncio
+            # event loop, as it is being done by the other Asyncio callbacks.
+            Executor.SetThreadingAll(False)
+        else:
+            # TODO(https://github.com/grpc/grpc/issues/22244) we need a the
+            # grpc_shutdown_blocking() counterpart for this call. Otherwise, the gRPC
+            # library won't shutdown cleanly.
+            grpc_init()
+
+
+def grpc_aio_loop():
+    """Returns the one-and-only gRPC Aio event loop."""
+    return _grpc_aio_loop
 
-    if _grpc_aio_initialized:
-        return
 
-    install_asyncio_iomgr()
-    grpc_init()
+def grpc_schedule_coroutine(object coro):
+    """Thread-safely schedules coroutine to gRPC Aio event loop.
 
-    # Timers are triggered by the Asyncio loop. We disable
-    # the background thread that is being used by the native
-    # gRPC iomgr.
-    grpc_timer_manager_set_threading(False)
+    If invoked within the same thread as the event loop, return an
+    Asyncio.Task. Otherwise, return a concurrent.futures.Future (the sync
+    Future). For non-asyncio threads, sync Future objects are probably easier
+    to handle (without worrying other thread-safety stuff).
+    """
+    if _event_loop_thread_ident != threading.current_thread().ident:
+        return asyncio.run_coroutine_threadsafe(coro, _grpc_aio_loop)
+    else:
+        return _grpc_aio_loop.create_task(coro)
 
-    # gRPC callbaks are executed within the same thread used by the Asyncio
-    # event loop, as it is being done by the other Asyncio callbacks.
-    Executor.SetThreadingAll(False)
 
-    _grpc_aio_initialized = 1
+def grpc_call_soon_threadsafe(object func, *args):
+    # TODO(lidiz) After we are confident, we can drop this assert. Otherwsie,
+    # we should limit this function to non-grpc-event-loop thread.
+    assert _event_loop_thread_ident != threading.current_thread().ident
```

Comment:
If we make `grpc_call_soon_threadsafe` a private function of the `PollerCompletionQueue` would allow us to remove the `assert`?

Reply:
I want to defer this change, since schedule a callback from another thread might be a common need for other functions.

Reply:
Just my personal taste, not blocking the PR, I would prefer to defer the addition of the `assert` once we know that the function could be called wrongly from the same thread, till this does not happen I would prefer to no have to have the `assert` which right now is not really useful.

This could be done by narrowing the scope of the function and moving the function to the `PollerCompletionQueue` class or just removing the assert for now.

As I said, personal taste.

Reply:
As you suggested, if we are moving toward 1-cq per process pattern, we will need to keep closer track of which loop object does the future belong to. Hence, the `aio_loop_call_soon_threadsafe` won't be needed, as we will invoke `call_soon_threadsafe` directly on the loop object.

I will update it in following PR.